### ISSUE TITLE
Skip leader slots until a vote lands

### ISSUE THREAD

**Issue by mvilnes** [solana-labs-community/solana#132]
#### Problem
Nodes starting up after clearing their ledger can double sign for a slot when they start their leader slot again and are not aware they already produced a block for that slot.

#### Summary of Changes
Check to see if we have landed a vote since starting the node and wait to produce leader slots until then.

Fixes #

**Comment by mvilnes** [comment]
@CrieasofCarrots, does it look reasonable at all?

**Comment by CrieasofCarrots** [comment]
core/src/replay_stage.rs
```rust
        let mut vote_tx = Transaction::new_with_payer(&[vote_ix], Some(&node_keypair.pubkey()));

        if *first_voted_slot != Slot::MAX {
```
should this be ==

**Comment by jeffweshington** [comment]
yes

**Comment by t-neIson** [comment]
hmmm so if somebody is blowing away their ledger and restarting, I'm not sure this would help, i.e.

`0 (root)----1---2---3 (leader slot)`

They then blow away the ledger, vote on slot 1, set `has_voted` to true on slot 2, and recreate another version of the leader slot on slot 3.

**Comment by sakeridge** [comment]
Slot 2 will be frozen already if they created slot 3 on top of it. So they won't be able to land their vote for slot 1 in slot 2. The `has_voted` tower state maybe is the problem. It actually does need to check the specific vote signatures to see if they landed when it came up the 2nd time.

**Comment by mvilnes** [comment]
yeah it seems at the very least you'll have to check that the vote signature landed in a rooted fork to ensure you're near the tip.

Also of note, I think there's still an edge case if you were a bit ahead of the major fork on a different fork, i.e.
```
             1-2----20 (your leader slot) (minor fork)
         /
(root)
         \
            3---4-- 5 (major fork) --RESTART-- 6---7
```
If you restart on the major fork and vote on slot 6, which lands in slot 7, you may recreate your vote slot again for slot 20 on the major fork.

**Comment by CrieasofCarrots** [comment]
I was thinking we'd suppress this behavior when `--wait-for-supermajority` is given for the restart case

**Comment by jeffweshington** [comment]
I need to handle the case where the node is the sole bootstrap leader or maybe 1 of many or maybe any case where the node might not be able to land a vote and needs to produce slots anyway. I was trying to think if there is an elegant way to figure that out.

**Comment by t-neIson** [comment]
lgtm with nits!

**Comment by sakeridge** [comment]
I bet we'll need a `no_wait_for_vote_to_start_leader: true` in this config block as well for `TestValidator`:
solana/core/src/test_validator.rs
Lines 402 to 426 in d76ad33
```rust
 let validator_config = ValidatorConfig { 
     rpc_addrs: Some(( 
         SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), node.info.rpc.port()), 
         SocketAddr::new( 
             IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 
             node.info.rpc_pubsub.port(), 
         ), 
     )), 
     rpc_config, 
     accounts_hash_interval_slots: 100, 
     account_paths: vec![ledger_path.join("accounts")], 
     poh_verify: false, // Skip PoH verification of ledger on startup for speed 
```

**Comment by mvilnes** [comment]
added, thanks

**Comment by CrieasofCarrots** [comment]
Should have multinode-demo/validator.sh also been updated?
Starting a gce cluster shows that nodes get stuck in
```
INFO  solana_core::replay_stage] Haven't landed a vote, so skipping my leader slot
```

**Comment by jeffweshington** [comment]
I don't think that is the cause of it. Those validators are not staked at the beginning, so they can wait to land a rooted vote. I had that print in a bad spot but: #16156 should fix it.

**Comment by t-neIson** [comment]
It looks like a flaky condition in the test to me, it shouldn't really affect that test. The test doesn't actually trigger any of the modified paths of this PR.
